sdcm/sct_runner.py

```diff
@@ -930,39 +1040,111 @@ def list_sct_runners(test_status: Optional[str] = None, test_runner_ip: str = No
+                timeout_flag, sct_runner_info.logs_collected)
+
+    if test_status == "SUCCESS" and sct_runner_info.logs_collected:
+        sct_runner_info.sct_runner_class.set_tags(sct_runner_info, {"keep": "0", "keep-action": "terminate"})
+        sct_runner_info.keep = 0
+        sct_runner_info.keep_action = "terminate"
+        return sct_runner_info
+
+    if not timeout_flag and sct_runner_info.logs_collected:
+        current_run_time_hrs = int((utc_now - sct_runner_info.launch_time).total_seconds() // 3600)
+        new_keep_value = int(sct_runner_info.keep) - current_run_time_hrs + 6
+
+        if new_keep_value > 0:
+            sct_runner_info.sct_runner_class.set_tags(sct_runner_info, {"keep": str(new_keep_value)})
+            sct_runner_info.keep = new_keep_value
+        return sct_runner_info
+
+    if test_status is not None and test_status != "RUNNING" and not sct_runner_info.logs_collected:
+        sct_runner_info.sct_runner_class.set_tags(sct_runner_info, {"keep": "alive", "keep_action": "keep"})
+        sct_runner_info.keep = "alive"
+        sct_runner_info.keep_action = "keep"
+        return sct_runner_info
+
+    LOGGER.info("No changes to make to runner tags.")
+    return sct_runner_info
+
+
+def clean_sct_runners(test_status: str,
+                      test_runner_ip: str = None,
+                      dry_run: bool = False) -> None:
+    sct_runners_list = list_sct_runners(test_runner_ip=test_runner_ip)
+    timeout_flag = False
+    runners_terminated = 0
+    end_message = ""
+
+    for sct_runner_info in sct_runners_list:
+        LOGGER.info("Attempting to create runner remoter with host: %s, region_name: %s",
+                    test_runner_ip, sct_runner_info.region_az)
+        cmd = 'cat /home/ubuntu/sct-results/latest/events_log/critical.log | grep "TestTimeoutEvent"'
+        if sct_runner_info.cloud_provider == "aws":
+            sct_runner_name = sct_runner_info.instance_name.split(" ")[0]
+        else:
+            sct_runner_name = sct_runner_info.instance_name
+        timeout_flag = bool(ssh_run_cmd(command=cmd, test_id=sct_runner_info.test_id,
+                                        node_name=sct_runner_name).stdout)
+        utc_now = datetime.datetime.now(tz=datetime.timezone.utc)
+        LOGGER.info("UTC now: %s", utc_now)
+
+        sct_runner_info = _manage_runner_keep_tag_value(test_status=test_status, utc_now=utc_now,
+                                                        timeout_flag=timeout_flag, sct_runner_info=sct_runner_info)
         if sct_runner_info.keep:
-            if "alive" in sct_runner_info.keep:
-                LOGGER.info("Skip %s because `keep' == `alive'", sct_runner_info)
-                continue
+            if "alive" in str(sct_runner_info.keep):
+                LOGGER.info("Skip %s because `keep' == `alive. No runners have been terminated'", sct_runner_info)
+                return
```

Comment:
when doing `return` we stop after first node. Is it intended?

Reply:
we should have only a single SCT Runner, no?

Reply:
We have a `hydra-clean-runner-instances.jenkinsfile` that uses the `clean-runner-instances` command without providing any runner ips. I'm not sure this is really needed (and it's a bit of playing with fire), but I didn't want to remove that functionality. In most cases though we'll be using this by providing a runner ip and test status, so only one runner will be managed.

As for the `continue` vs `return`, it's fixed.